package NYT.NChunkClient.NProto;

import "yt/ytlib/chunk_client/proto/chunk_service.proto";

import "yt/yt/server/lib/chunk_server/proto/job_common.proto";

import "yt_proto/yt/core/misc/proto/error.proto";
import "yt_proto/yt/core/misc/proto/guid.proto";

import "yt_proto/yt/client/node_tracker_client/proto/node_directory.proto";
import "yt_proto/yt/client/table_chunk_format/proto/chunk_meta.proto";

////////////////////////////////////////////////////////////////////////////////

message TReplicateChunkJobSpecExt
{
    extend NChunkServer.NProto.TJobSpec
    {
        optional TReplicateChunkJobSpecExt replicate_chunk_job_spec_ext = 201;
    }

    reserved 5;

    required NYT.NNodeTrackerClient.NProto.TNodeDirectory node_directory = 1;

    // Encoded chunk id.
    required NYT.NProto.TGuid chunk_id = 3;

    // Medium where the chunk replica resides.
    required int32 source_medium_index = 4;

    // Nodes where new replicas must be created.
    // Replica indexes are not used.
    // Medium indexes indicate destination medium.
    repeated fixed64 target_replicas = 6;

    // True if the chunk is consistently placed and this job was scheduled through pull replication.
    optional bool is_pull_replication_job = 7;
}

////////////////////////////////////////////////////////////////////////////////

message TRemoveChunkJobSpecExt
{
    extend NChunkServer.NProto.TJobSpec
    {
        optional TRemoveChunkJobSpecExt remove_chunk_job_spec_ext = 204;
    }

    // Encoded chunk id.
    required NYT.NProto.TGuid chunk_id = 1;

    // Medium where the chunk replica resides.
    // COMPAT(aleksandra-zh)
    optional int32 medium_index = 2;

    // If true and the chunk is actually missing at node, the job is considered successful
    // rather than failed.
    optional bool chunk_is_dead = 5;

    optional NYT.NProto.TGuid location_uuid = 6;

    reserved 3, 4;
}

////////////////////////////////////////////////////////////////////////////////

message TRepairChunkJobSpecExt
{
    extend NChunkServer.NProto.TJobSpec
    {
        optional TRepairChunkJobSpecExt repair_chunk_job_spec_ext = 202;
    }

    reserved 5;

    required int32 erasure_codec = 1;

    required NYT.NNodeTrackerClient.NProto.TNodeDirectory node_directory = 2;

    // Unencoded chunk id.
    required NYT.NProto.TGuid chunk_id = 3;

    // Nodes where existing replicas reside.
    // Replica indexes indicate specific chunk parts.
    repeated fixed64 source_replicas = 10;

    // Nodes where repaired replicas must be created.
    // Replica indexes indicate specific chunk parts to be repaired.
    // Medium indexes are same and indicate destination medium.
    repeated fixed64 target_replicas = 7;

    // True, if this repair job was scheduled as a part of node decommission.
    optional bool decommission = 6;

    // Only given for erasure journals.
    optional int64 row_count = 8;

    // Is repairable chunk a striped erasure chunk?
    optional bool striped_erasure_chunk = 9;

    reserved 4;
}

////////////////////////////////////////////////////////////////////////////////

message TSealChunkJobSpecExt
{
    extend NChunkServer.NProto.TJobSpec
    {
        optional TSealChunkJobSpecExt seal_chunk_job_spec_ext = 203;
    }

    // The final number of rows this chunk must have.
    required int64 row_count = 1;

    required NYT.NNodeTrackerClient.NProto.TNodeDirectory node_directory = 2;

    // Unencoded chunk id.
    required NYT.NProto.TGuid chunk_id = 3;

    // Erasure codec.
    optional int32 codec_id = 6; // NErasure::ECodec

    // Medium where the chunk replica resides.
    required int32 medium_index = 4;

    // Nodes where existing replicas reside.
    // Replica indexes are not used.
    repeated fixed64 source_replicas = 7;

    reserved 5;
}

////////////////////////////////////////////////////////////////////////////////

message TMergeChunkInfo
{
    required NYT.NProto.TGuid id = 1;

    // Nodes where existing replicas reside.
    // Replica indexes are not used.
    repeated fixed64 source_replicas = 5;
    optional int64 row_count = 3 [default = 0];
    optional int32 erasure_codec = 4 [default = 0]; // NErasure::ECodec

    reserved 2;
}

message TChunkMergerWriterOptions
{
    optional NYT.NTableClient.NProto.TTableSchemaExt schema = 1;
    required int32 compression_codec = 2; // NCompression::ECodec
    optional int32 optimize_for = 3; // EOptimizeFor
    optional int32 chunk_format = 9; // EChunkFormat
    optional bool enable_skynet_sharing = 4;
    optional int32 erasure_codec = 5 [default = 0]; // NErasure::ECodec
    optional int32 merge_mode = 6 [default = 2]; // NChunkClient::EChunkMergerMode, default is EChunkMergerMode::Deep
    optional int32 max_heavy_columns = 7 [default = 0];
    optional int64 max_block_count = 8;
    optional NYT.NProto.TGuid schema_id = 10;
}

message TMergeChunksJobSpecExt
{
    extend NChunkServer.NProto.TJobSpec
    {
        optional TMergeChunksJobSpecExt merge_chunks_job_spec_ext = 208;
    }

    required uint32 cell_tag = 1;
    repeated TMergeChunkInfo input_chunks = 2;
    required NYT.NProto.TGuid output_chunk_id = 3;
    required int32 medium_index = 4;
    required NYT.NNodeTrackerClient.NProto.TNodeDirectory node_directory = 5;
    required TChunkMergerWriterOptions chunk_merger_writer_options = 6;
    repeated fixed64 target_replicas = 7;
    optional bool validate_shallow_merge = 8;
    optional bool validate_chunk_meta_extensions = 9;
}

message TMergeChunksJobResultExt
{
    extend NChunkServer.NProto.TJobResult
    {
        optional TMergeChunksJobResultExt merge_chunks_job_result_ext = 210;
    }

    optional bool deep_merge_fallback_occurred = 1;

    optional NYT.NProto.TError shallow_merge_validation_error = 2;
    optional NYT.NProto.TError chunk_meta_validation_error = 3;
}

////////////////////////////////////////////////////////////////////////////////

message TAutotomizeChunkJobSpecExt
{
    extend NChunkServer.NProto.TJobSpec
    {
        optional TAutotomizeChunkJobSpecExt autotomize_chunk_job_spec_ext = 209;
    }

    // Body chunk.
    required NYT.NProto.TGuid body_chunk_id = 1;
    required int64 body_chunk_first_overlayed_row_index = 2;
    repeated fixed64 body_chunk_replicas = 3;
    required int64 body_chunk_replica_lag_limit = 4;

    // Tail chunk.
    required NYT.NProto.TGuid tail_chunk_id = 5;

    // Journal parameters.
    required int32 read_quorum = 6;
    required int32 write_quorum = 7;
    required int32 medium_index = 8;
    required int32 erasure_codec = 9; // NErasure::ECodec
    required int32 replication_factor = 10;
    required bool overlayed = 11;
    required NYT.NNodeTrackerClient.NProto.TNodeDirectory node_directory = 12;
}

message TAutotomizeChunkJobResultExt
{
    extend NChunkServer.NProto.TJobResult
    {
        optional TAutotomizeChunkJobResultExt autotomize_chunk_job_result_ext = 201;
    }

    required NYT.NProto.TGuid body_chunk_id = 1;
    required NYT.NChunkClient.NProto.TChunkSealInfo body_chunk_seal_info = 2;
    required NYT.NProto.TGuid tail_chunk_id = 3;
    required NYT.NChunkClient.NProto.TChunkSealInfo tail_chunk_seal_info = 4;
}

////////////////////////////////////////////////////////////////////////////////

message TReincarnateChunkJobSpecExt
{
    extend NChunkServer.NProto.TJobSpec
    {
        optional TReincarnateChunkJobSpecExt reincarnate_chunk_job_spec_ext = 211;
    }

    required NYT.NProto.TGuid old_chunk_id = 1;
    required int32 erasure_codec = 2;
    required int32 compression_codec = 3;
    required NYT.NProto.TGuid new_chunk_id = 4;
    repeated fixed64 source_replicas = 10;
    repeated fixed64 target_replicas = 6;
    required int32 medium_index = 7;
    required NYT.NNodeTrackerClient.NProto.TNodeDirectory node_directory = 8;
    required bool enable_skynet_sharing = 9;
    required NYT.NProto.TGuid transaction_id = 11;

    reserved 5;
}

////////////////////////////////////////////////////////////////////////////////
